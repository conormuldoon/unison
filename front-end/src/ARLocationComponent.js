import PropTypes from 'prop-types';
import React, { useState } from 'react';
import './App.css';


/**
 * A component that enables locations for tracking to be either added or removed. 
 * The user name and password is posted to the server and the obtain data callback is executed if successful.
 * 
 * @component
 * 
 */
function ARLocationComponent({ createLocation, createRemove }) {

  const [displayAdd, setDisplayAdd] = useState(false);

  function toggleDisplayAdd() {

    setDisplayAdd(!displayAdd);
  }

  function hideAdd() {
    if (displayAdd)
      setDisplayAdd(false);
  }

  return (
    <div>
      <div className='pLeft'>
        {createLocation(toggleDisplayAdd, displayAdd, hideAdd)}
      </div>
      {createRemove && <div className='pLeft'>
        {createRemove(hideAdd)}
      </div>}
    </div>
  );
}


ARLocationComponent.propTypes = {


  createLocation: PropTypes.func.isRequired,
  createRemove: PropTypes.func,
}

export default ARLocationComponent;