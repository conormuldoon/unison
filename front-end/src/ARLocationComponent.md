ARLocationComponent example:

```js
<div style={{paddingLeft:200}}>
<ARLocationComponent location="UCD"  />
</div>
```