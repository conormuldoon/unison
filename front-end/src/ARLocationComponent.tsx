
import React, { useState } from 'react';
import './App.css';

import PropTypes from 'prop-types';

export interface ARLocationProps {

  /** A factory that creates a add location form. */
  createLocation: (toggleDisplayAdd: () => void, displayAdd: boolean, hideAdd: () => void) => React.ReactNode;

  /** A factory that creates a component from removing locations. */
  createRemove: ((hideAdd: () => void) => React.ReactNode) | null;
}

/**
 * A component that enables locations for tracking to be either added or removed. 
 * The user name and password is posted to the server and the obtain data callback is executed if successful.
 * 
 * @component
 * 
 */
function ARLocationComponent({ createLocation, createRemove }: ARLocationProps): JSX.Element {

  const [displayAdd, setDisplayAdd] = useState(false);

  function toggleDisplayAdd() {

    setDisplayAdd(!displayAdd);
  }

  function hideAdd() {
    if (displayAdd)
      setDisplayAdd(false);
  }

  return (
    <div>
      <div className='pLeft'>
        {createLocation(toggleDisplayAdd, displayAdd, hideAdd)}
      </div>
      {createRemove && <div className='pLeft'>
        {createRemove(hideAdd)}
      </div>}
    </div>
  );
}

ARLocationComponent.propTypes = {


  createLocation: PropTypes.func.isRequired,
  createRemove: PropTypes.func,
}


export default ARLocationComponent;