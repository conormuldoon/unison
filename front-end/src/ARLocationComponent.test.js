import "@testing-library/jest-dom/extend-expect";
import React from 'react';
import ReactDOM from 'react-dom';
import { fireEvent, render } from "react-testing-library";
import ARLocationComponent from './ARLocationComponent';

it('renders without crashing', async () => {

  const div = document.createElement('div');

  ReactDOM.render(<ARLocationComponent obtainData={() => { }} />, div);
  ReactDOM.unmountComponentAtNode(div);
});


it('mathes snapshot', () => {


  const { container } = render(<ARLocationComponent obtainData={() => { }} />);

  expect(container).toMatchSnapshot();

});

const linksProperty = {
  "name": "UCD",
  "_links": {
    "self": {
      "href": "http://localhost:8080/locationCollection/UCD"
    },
    "cloudiness": {
      "href": "http://localhost:8080/locationCollection/UCD/cloudiness{?fromDate,toDate}",
      "templated": true
    },
    "cloudLevel": {
      "href": "http://localhost:8080/locationCollection/UCD/cloudLevel{?fromDate,toDate}",
      "templated": true
    },
    "dewPoint": {
      "href": "http://localhost:8080/locationCollection/UCD/dewPoint{?fromDate,toDate}",
      "templated": true
    },
    "humidity": {
      "href": "http://localhost:8080/locationCollection/UCD/humidity{?fromDate,toDate}",
      "templated": true
    },
    "precipitation": {
      "href": "http://localhost:8080/locationCollection/UCD/precipitation{?fromDate,toDate}",
      "templated": true
    },
    "pressure": {
      "href": "http://localhost:8080/locationCollection/UCD/pressure{?fromDate,toDate}",
      "templated": true
    },
    "temperature": {
      "href": "http://localhost:8080/locationCollection/UCD/temperature{?fromDate,toDate}",
      "templated": true
    },
    "windDirection": {
      "href": "http://localhost:8080/locationCollection/UCD/windDirection{?fromDate,toDate}",
      "templated": true
    },
    "windSpeed": {
      "href": "http://localhost:8080/locationCollection/UCD/windSpeed{?fromDate,toDate}",
      "templated": true
    }
  }
}


it('toggles add location correctly', () => {
  const confirmSpy = jest.spyOn(window, 'confirm');
  confirmSpy.mockImplementation(() => true);

  const location = { name: 'UCD', links: { self: '/location/UCD' } };
  const { getByTestId, getByText } = render(<ARLocationComponent location={location} obtainData={() => { }} linksProperty={linksProperty} />);
  expect(getByTestId('lf-button')).toHaveTextContent('Add Location');
  expect(getByTestId('rm-button')).toHaveTextContent('Remove ' + location.name);
  fireEvent.click(getByText('Add Location'));
  expect(getByTestId('lf-button')).toHaveTextContent('Hide');
  fireEvent.click(getByText('Hide'));
  expect(getByTestId('lf-button')).toHaveTextContent('Add Location');
  fireEvent.click(getByText('Add Location'));
  fireEvent.click(getByText('Remove ' + location.name));
  expect(getByTestId('lf-button')).toHaveTextContent('Add Location');

  confirmSpy.mockClear();

});

