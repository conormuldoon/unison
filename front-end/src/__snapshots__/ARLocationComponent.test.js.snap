// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <div
      class="pLeft"
    >
      <div
        id="pLeft"
      >
        <button
          data-testid="lf-button"
        >
          Add Location
        </button>
      </div>
    </div>
  </div>
</div>
`;
